### PR TITLE
refactors the three responsibilities of equal/hashcode/signature

### PR DESCRIPTION
**Pull request by monperrus**
Before:
The three responsibilities of equals/hashCode/signature were handled by the same code in SignaturePrinter,
with two major problems:
- no deep equality for package, class and method
- low performance of hashCode

Now:
The three responsibilities are handled by three different visitors:
- SignaturePrinter for signatures
- EqualVisitor for equal
- HashcodeVisitor for equal

### CODE REVIEW

**Comment by GerardPaligot** [comment]
PR launched on our Jenkins. Results soon.

**Comment by monperrus** [comment]
fingers crossed :-)

**Comment by monperrus** [comment]
we also have to find a solution for the Log.isDebug in equal.

**Comment by GerardPaligot** [comment]
Faster solution but worst: isRunningTests()
Lower solution but better: Specific logger for this check. By default, disabled but we can enable it for some tests like `MainTest`.

**Comment by GerardPaligot** [comment]
Jenkins ok.

**Comment by GerardPaligot** [comment]
Jenkins not ok with Google Guice: http://193.51.236.32:8080/job/Google%20Guice/76/

File: src/main/java/spoon/support/visitor/EqualVisitor.java [hunk 1]
```diff
@@ -0,0 +1,700 @@
...
+			for (CtTypeReference<?> b : ref.getBounds()) {
+				scan(b);
+				write(", ");
+			}
+		}
+	}
+
+	@Override
+	public <T> void visitCtTypeReference(CtTypeReference<T> reference) {
+		write(reference.getSimpleName());
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/visitor/EqualVisitor.java]: This should be `reference.getQualifiedName()` (fix Google Guice).

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 2]
```diff
@@ -158,23 +161,39 @@ public CtElementImpl() {
...
-		String other = o.getSignature();
+		String current = getDeepRepresentation(this);
+		String other = getDeepRepresentation(o);
 		if (current.length() <= 0 || other.length() <= 0) {
 			throw new ClassCastException("Unable to compare elements");
 		}
 		return current.compareTo(other);
 	}
 
+	public String getDeepRepresentation(CtElement elem) {
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: This doesn't work. e.g., when you'll build the deep representation of a method like this one:

```java
public void toProvider(Provider<? extends T> provider) {
}
```

The result of `getDeepRepresentation` will be: `void toProvider(javax.inject.Provider,)voidprovider{}`.

**Comment by monperrus** [comment]
thanks for the debug. updated.

**Comment by monperrus** [comment]
> Lower solution but better: Specific logger for this check. By default,
> disabled but we can enable it for some tests like |MainTest|.
I like this solution. But, it should not be for MainTest but for all tests.

Can we have two different log configurations, for testing and for deploying?

If yes, we have the solution!

**Comment by GerardPaligot** [comment]
Ok with Jenkins.

**Pull request merged**